File: client/python/tests/pages/contact_list_page.py
"""Page Object — Contacts app: contact list screen."""

from __future__ import annotations

from maestro_runner.models import ExecutionResult

from tests.pages import BasePage
from tests.pages.edit_contact_page import EditContactPage


class ContactListPage(BasePage):
    """The main Contacts list screen."""

    APP_ID = "com.google.android.contacts"

    def launch(self, clear_state: bool = True) -> ExecutionResult:
        result = self.client.launch_app(self.APP_ID, clear_state=clear_state)
        self.wait_for_animation()
        return result

    def open_create_contact(self) -> EditContactPage:

        self.client.tap(text="Create contact|Add contact|New contact")
        self.wait_for_animation()
        return EditContactPage(self.client)

    def assert_contact_visible(self, name: str) -> ExecutionResult:
        parts = name.split()
        if len(parts) == 2:
            first, last = parts
            pattern = f".*{first}.*{last}.*|.*{last}, {first}.*"
            return self.client.assert_visible(text_pattern=pattern)
        return self.client.assert_visible(text=name)

File: client/typescript/tests/pages/ContactListPage.ts
/** Page Object — Contacts app: contact list screen. */

import { MaestroClient, ExecutionResult } from "../../src";
import { BasePage } from "./BasePage";
import { EditContactPage } from "./EditContactPage";

export class ContactListPage extends BasePage {
  static readonly APP_ID = "com.google.android.contacts";

  constructor(client: MaestroClient) {
    super(client);
  }

  async launch(clearState: boolean = true): Promise<ExecutionResult> {
    const result = await this.client.launchApp(ContactListPage.APP_ID, { clearState });
    await this.waitForAnimation();
    return result;
  }

  async openCreateContact(): Promise<EditContactPage> {
    await this.client.tap({ text: "Create contact|Add contact|New contact" });
    await this.waitForAnimation();
    return new EditContactPage(this.client);
  }

  async assertContactVisible(name: string, timeoutMs?: number): Promise<ExecutionResult> {
    const parts = name.trim().split(/\s+/);
    if (parts.length === 2) {
      const [first, last] = parts;
      const pattern = `.*${first}.*${last}.*|.*${last}, ${first}.*`;
      return this.client.assertVisible({ textPattern: pattern, timeoutMs });
    }
    return this.client.assertVisible({ text: name, timeoutMs });
  }
}
